### PR TITLE
Move from LGTM to GitHub code scanning (step 1: Python)

### PR DESCRIPTION
**Pull request by martinpitt**
LGTM is deprecated and replaced with CodeQL, see
https://github.blog/2022-08-15-the-next-step-for-lgtm-com-github-code-scanning/
    
A full scan with Python, JS, and C generates some 180 alerts, 40 of
which are "High".  So let's start small, and enable Python only.

----

This also contains a batch of fixes, to address most  Python alerts. I dismissed some of them which were false positives, so that the list is mostly clean now. The alerts can be seen in the "Code scanning results / CodeQL" status check, and also here: https://github.com/cockpit-project/cockpit/security/code-scanning?query=is%3Aopen+branch%3Acodeql

I set the "Check failure" in the [scan settings](https://github.com/cockpit-project/cockpit/settings/security_analysis) to a very high value of "high or higher/errors and warnings", as wading through all issues will take some time. We can gradually lower it over time. (e.g. in bots we now let it fail on all notices even).

I disabled the LGTM.com GitHub App already. It's going away soon, and nobody ever looked at its results anyway.

### CODE REVIEW

File: tools/urls-check
Comment (martinpitt) [reply tools/urls-check:103]: Hmm, that's what I got for the *old* code, and the new code was supposed to fix that.

Comment (martinpitt) [reply tools/urls-check]: Fixed.

**Comment by martinpitt** [comment]
Okay, I got a green run. For some reason, https://github.com/cockpit-project/cockpit/security/code-scanning?query=pr%3A17916+tool%3ACodeQL+is%3Aopen still shows 5 open issues, but they are all fixed here -- and clicking on them indeed shows "fixed in branch...". I suppose that'll clear up once this lands in main, that's just some weird UI quirk.

File: test/common/pixeldiff.html
Comment (allisonkarlitskaya) [inline test/common/pixeldiff.html:419]: hihi js?

File: pkg/storaged/luksmeta-monitor-hack.py
Comment (allisonkarlitskaya) [inline pkg/storaged/luksmeta-monitor-hack.py:45]: I think I'd prefer

```python
clevis = header_object.get('clevis')
if clevis is None:
    return None
```
...followed by the rest of the stuff, unindented

File: test/common/lcov.py
Comment (allisonkarlitskaya) [inline test/common/lcov.py:177]: Same with the comment above.  This is effectively the "error case" and should be treated like that, instead of indenting the "normal case".

File: tools/make-compile-commands
Comment (allisonkarlitskaya) [inline tools/make-compile-commands:47]: I think I had that there as documentation.

Can you add a comment for this one, as per the `# match group 4` line just below?

File: pkg/storaged/mount-users.py
Comment (allisonkarlitskaya) [inline pkg/storaged/mount-users.py:25]: This screams out for `shlex.join()` or `map(shlex.quote, ...)`.

File: pkg/storaged/mount-users.py
Comment (allisonkarlitskaya) [inline pkg/storaged/mount-users.py:24]: Minor nag: there could be non-utf8 stuff in there, right?  What happens in that case?

File: pkg/storaged/nfs-mounts.py
Comment (allisonkarlitskaya) [inline pkg/storaged/nfs-mounts.py:49]: Or more idiomatically:

```
    for line in f:
```

File: pkg/storaged/nfs-mounts.py
Comment (allisonkarlitskaya) [inline pkg/storaged/nfs-mounts.py:70]: ...and that one is (approximately)

`   lines = f.readlines()`

Since you strip the lines in the loop below, it ought to be equivalent.

File: pkg/storaged/nfs-mounts.py
Comment (allisonkarlitskaya) [inline pkg/storaged/nfs-mounts.py:71]: ...`lines[-1]`...

Comment (allisonkarlitskaya) [reply pkg/storaged/nfs-mounts.py:71]: I mean, there's an awful lot that could be changed here, actually.  Perhaps it's out of scope for this PR.

File: test/verify/check-system-services
Comment (allisonkarlitskaya) [inline test/verify/check-system-services:130]: https://docs.python.org/3/library/stdtypes.html#str.removesuffix

> New in version 3.9.


gr.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
Looks great, thanks.

Of all of the possible choices for what to call the damn thing that `open()` returns, you settled on `f`?  seriously?  OK :shrug:

**Comment by martinpitt** [comment]
> Of all of the possible choices for what to call the damn thing that open() returns, you settled on f?

Hmm, `git grep 'open.*as f:'| wc -l` has 26 hits on main. We have 4 `fd`, 4 `file`, 4 `fp`, and then a long tail of other names. I just went with the majority, but if there's a different convention in the python world, I'm happy to adjust :grin:

File: tools/urls-check
Comment (martinpitt) [reply tools/urls-check:103]: No, there aren't; I ran urls-check (with a hack to not touch GitHub), and it's happy.

File: pkg/storaged/luksmeta-monitor-hack.py
Comment (martinpitt) [reply pkg/storaged/luksmeta-monitor-hack.py:45]: Done.

File: tools/make-compile-commands
Comment (martinpitt) [reply tools/make-compile-commands:47]: Good idea, done.

File: pkg/storaged/nfs-mounts.py
Comment (martinpitt) [reply pkg/storaged/nfs-mounts.py:70]: Done.

File: test/verify/check-system-services
Comment (martinpitt) [reply test/verify/check-system-services:130]: :man_shrugging:

File: pkg/storaged/nfs-mounts.py
Comment (martinpitt) [reply pkg/storaged/nfs-mounts.py:71]: Right, but I at least did the simple `-1` change, that's obvious and safe.

File: pkg/storaged/mount-users.py
Comment (martinpitt) [reply pkg/storaged/mount-users.py:25]: Interestingly this does the  wrong thing with the trailing `\0`, as it ends up as an extra `''` argument. This works, though:

    shlex.join(f.read().rstrip("\0").split("\0"))

So going use that. Added as an extra commit as it's a behaviour change, and unrelated to opening files.

File: pkg/storaged/mount-users.py
Comment (martinpitt) [reply pkg/storaged/mount-users.py:24]: Too complex/too much diversion for me for the context of this PR. I'll split this out into a new issue and leave it to @mvollmer.

Comment (martinpitt) [reply pkg/storaged/mount-users.py:24]: #17920

File: test/verify/check-testlib
Comment (martinpitt) [reply test/verify/check-testlib:174]: Agreed in principle, but too much blockage/diversion for this big PR, I dropped the commit. Let's first add the bots PR, land it, and then use it consistently. Or check if it's really worth doing a new bots API, TBH I'd just use a shell command here. No time to think about it right now.

**Comment by martinpitt** [comment]
[bwah old Python on RHEL 8](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221115-153416-edbc7caf-rhel-8-7-distropkg/TestStorageMounting-testMounting-rhel-8-7-distropkg-127.0.0.2-3001-FAIL.png).. No `shlex.join` for us. I drop that commit again.

File: test/verify/check-testlib
Comment (martinpitt) [reply test/verify/check-testlib]: I reworked this as you proposed. I wasted way too much time on figuring out the weird test failure because I forgot `O_TRUNC`.. :angry:

File: pkg/users/account-create-dialog.js
Comment (cockpituous) [inline pkg/users/account-create-dialog.js:219]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221115-164846-1bf8b990-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/password-dialogs.js
Comment (cockpituous) [inline pkg/users/password-dialogs.js:197]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221115-164846-1bf8b990-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: test/verify/check-testlib
Comment (martinpitt) [reply test/verify/check-testlib:134]: This is stupid now, I waived that alert. Still better than what we had before.

**Comment by martinpitt** [comment]
@allisonkarlitskaya This is hard to get green, due to the 3x affected retries. But should in principle be good now, PTAL?

**Comment by martinpitt** [comment]
Oops, this [insights failure](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221116-052939-1bf8b990-rhel-8-7/log.html) is real, debugging.

File: pkg/users/account-create-dialog.js
Comment (cockpituous) [inline pkg/users/account-create-dialog.js:219]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221116-052945-1bf8b990-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/password-dialogs.js
Comment (cockpituous) [inline pkg/users/password-dialogs.js:197]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221116-052945-1bf8b990-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: test/verify/check-testlib
Comment (allisonkarlitskaya) [inline test/verify/check-testlib:135]: This isn't necessary.  If `open()` fails, it'll raise an exception.

File: test/verify/check-testlib
Comment (allisonkarlitskaya) [inline test/verify/check-testlib:137]: If you wanted to be very very strict, though, you might decide you want to deal with the case where `write()` fails (`ENOSPC`?) and then you'll leak the `fd`...

I like to use `finally:` blocks when dealing with fds.

```python
try:
    os.write(fd, content.encode())
finally:
    os.close(fd)
```

**Review by allisonkarlitskaya — DISMISSED** [review]
I'd be happy with landing this as it is.  It's pretty big...

If you want to make the tweaks suggested, please do so.  Otherwise, ignore them, or follow up :)

File: pkg/users/account-create-dialog.js
Comment (cockpituous) [inline pkg/users/account-create-dialog.js:219]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221116-081027-9b3960eb-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/password-dialogs.js
Comment (cockpituous) [inline pkg/users/password-dialogs.js:197]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221116-081027-9b3960eb-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: test/verify/check-testlib
Comment (martinpitt) [reply test/verify/check-testlib]: Good idea. I also strengthened the os.write() to assert a full write. It's "just" a test, so I think we can affort to be stricter than necessary, and not waffle around with retry loops and such.

File: test/verify/check-testlib
Comment (martinpitt) [reply test/verify/check-testlib:135]: Waived, this is a false positive.

**Comment by allisonkarlitskaya** [comment]
> [bwah old Python on RHEL 8](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221115-153416-edbc7caf-rhel-8-7-distropkg/TestStorageMounting-testMounting-rhel-8-7-distropkg-127.0.0.2-3001-FAIL.png).. No `shlex.join` for us. I drop that commit again.

I had written that this was new in 3.8 in my original review comment, but GitHub ate it.  My second attempt was much more terse (but that's why I mentioned `map(shlex.quote, ...)` as the alternative, indeed.

File: pkg/users/account-create-dialog.js
Comment (cockpituous) [inline pkg/users/account-create-dialog.js:219]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221116-122443-bc0875ab-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/password-dialogs.js
Comment (cockpituous) [inline pkg/users/password-dialogs.js:197]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17916-20221116-122443-bc0875ab-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: test/verify/check-testlib
Comment (allisonkarlitskaya) [inline test/verify/check-testlib:138]: lol.  hardcore.

**Review by allisonkarlitskaya — APPROVED** [review]
Looks good!  Sorry this one slipped off my radar for the entire day :rofl:

**Pull request merged**
